Collapse sample point indicators into dots if they cannot be displayed in full

OliBomby opened the pull request:
This toggle reduces visual clutter for mappers who prefer to not see hitsound samples while mapping.

OliBomby committed:
Add view menu toggle for sample points

bdach commented:
I guess I don't have a particularly strong objection to this in isolation, although I do have a few doubts:

- This is the only place in the entirety of editor where it is possible to set hitsounds without hotkeys, so hiding it is a bit painful. That said, given it defaults to off, I'd think anyone who did disable it for themselves should be able to show it again.
- Why do sample pieces specifically get this treatment but slider velocity pieces don't?

Probably need to gather more opinions on this, I'm not super sure I'd be fine with merging without a consult.

peppy commented:
Dunno, next are we adding velocity toggles? Then hitobject toggles?

I feel this is misdirected UX. Where's the request coming from? I'd sooner focus on making the sample editor and then considering making the display here less obtrusive to the point people aren't asking(?) for toggles.

OliBomby commented:
> Where's the request coming from?

I got the idea from a small discord discussion. I also heard other people wish to turn off the sample points. Clearly here the sample points add the most visual clutter, having the most objects in the timeline and easily overlapping neighbouring sample points. I think it then makes sense to have a toggle to turn them off because a lot of people map without doing hitsounds. For them it only adds noise.
![Discord_KztYdKyZ7z](https://github.com/user-attachments/assets/0b3c8677-d6ad-4837-8f0f-5a8222d476a9)

I'm also PR'ing this assuming there will be a sample editor in the future where you can better view the hitsounds in a separate tab. In that case it still makes sense to have this toggle, because there are two kinds of mappers:
- Mappers who map first, then hitsound: Map in compose with sample points toggled off, then once the map is done hitsound in the sample editor or toggle on the sample points to do simple hitsounds in compose.
- Mappers who map and hitsound at the same time: They map with sample points toggled on all the time.

> Dunno, next are we adding velocity toggles? Then hitobject toggles?

Please dont extrapolate this to the extreme. I can see velocity toggles maybe being added, but its really not necessary because velocity points are not nearly as common as sample points so they dont add much visual clutter. Hitobject toggle? not at all.

peppy commented:
I think a better initial approach may be to hide them when they begin to overlay? ie. require the user zoom in before it starts to display.

OliBomby commented:
> This looks like a rather extreme example? Looks like the timeline is zoomed out to the maximum and not even the hitobject displays are all that useful at this density? Also I dunno what the "so many colors" comment is supposed to be given that 6 of the 8 unique colours shown are _combo_ colours? Are we going to add a toggle for those too?

Maybe not a great example, but my point is that people want to toggle off the sample points. The visual clutter is there at normal zoom levels too.
![osu Game Rulesets Osu Tests_wWMOb7SuIF](https://github.com/user-attachments/assets/c08d6cb3-9d32-49c8-83b4-b72f2a958af3)
^ Information overload
![osu Game Rulesets Osu Tests_UlLfVPMAFG](https://github.com/user-attachments/assets/b0f62e9e-7ce9-41ef-9d07-57dc45cc1fe8)
^ Mindful and demure

> I think a better initial approach may be to hide them when they begin to overlay? ie. require the user zoom in before it starts to display.

This is an even weirder suggestion to me. The visibility should not be affected by the zoom level, that will get annoying very quick.

The problem is the information density of the sample points being much higher than the other elements in the timeline while some mappers dont even need this information, regardless of zoom level.

peppy commented:
> The visibility should not be affected by the zoom level, that will get annoying very quick. 

Then have them reduce to a small pink dot, rather than overlapping pink-huge-blobs-with-text-that's-not-legible-anyway. The idea is to keep definition as much as possible.

kadambishreyas commented:
agree with peppy for both sample and velocity honestly.
the clutter is caused by the size of each button + number of buttons, so minimizing to a circle (perhaps with a generous clickable area) to provide a clickable to toggle settings when there is no change in the visible statistics (i.e. sv for green labels, sampleset OR hitsound volume for pink labels) from the previous object seems like a massive improvement imo.

OliBomby committed:
Reduce sample point pieces to pink dot if zoomed out

bdach commented:
The size reduction seems to work ok, but why is the toggle still there?

https://github.com/user-attachments/assets/32c89d2b-cd48-4497-9eb8-5de96fecb360

I thought the goal was to have the dots shrink *instead of* the toggle, rather than *in addition to*? @peppy did I misunderstand?

peppy commented:
I'd prefer not having the toggle at all.

peppy committed:
Use bindable flow for zoom handling

peppy committed:
Make points slightly smaller when contracted and fix width getting obliterated

OliBomby commented:
> Seems to contract way too early depending on BPM/SV

The best point to contract depends on the density of the beatmap which various wildly between different BPMs and star ratings. I think we need a more sophisticated criterion to determine when to contract the sample points which looks at the hitobjects in the timeline.

peppy commented:
I think something could be done in `TimelineBlueprintContainer` to that efffect. Basically constantly watching for the lowest/highest SV of visible objects and adjusting a `BindableBool` governing the contracted state.

OliBomby committed:
contract sample point pieces based on smallest gap on the timeline

It finds the smallest distance between two sample point pieces on the alive timeline blueprints

OliBomby commented:
I made it contract the sample point pieces based on the sample point pieces which are closest together on the visible timeline, so basically it tried to prevent any overlap from happening and it will not contract if there is no overlap.

It's a lot better but ideally I'd want to mix contracted and non-contracted sample point pieces on the same timeline, so you don't have all the sample points contracting together when only one is overlapping. However I don't know how to get a list of just the sample point drawables on the timeline because of the way they exist as children of `TimelineHitObjectBlueprint`. It might require a refactor to put all the sample pieces on a seperate timeline part.

bdach commented:
> I made it contract the sample point pieces based on the sample point pieces which are closest together on the visible timeline, so basically it tried to prevent any overlap from happening and it will not contract if there is no overlap.

Doesn't work properly when seeking:

https://github.com/user-attachments/assets/e0cddb5e-5ff4-4b92-b84a-338c7f2a59ce

bdach reviewed (changes requested):
see above

OliBomby committed:
Use FinishTransforms to init visual state

OliBomby committed:
Update sample point contracted state each frame

OliBomby committed:
dont contract samples for stacked objects

OliBomby commented:
To fix the contraction not updating I followed the example of `updateStacking()` and made it update every frame. There doesn't seem to be an easy way to detect for changes in the `SelectionBlueprints`.

bdach commented on osu.Game/Screens/Edit/Compose/Components/Timeline/TimelineBlueprintContainer.cs:
```diff
@@ -118,9 +120,44 @@ protected override void Update()
+            double lastTime = double.PositiveInfinity;
+
+            // The blueprints are ordered in reverse chronological order
+            foreach (var selectionBlueprint in SelectionBlueprints)
```
This can't be done. While `SelectionBlueprints` will only contain the visible blueprints in the general case... it will also contain more blueprints if they are selected. In fact if you select every object you may see these contract for no reason:

https://github.com/user-attachments/assets/5cff6f47-85d0-4e70-99fb-e0141c02de2f

I imagine this will also make selecting all objects even slower than it usually is, although crude profiling shows maybe not very much:

![1729505224](https://github.com/user-attachments/assets/7fda5bf6-1768-4f5f-9fd0-b8aa5481f4f9)

OliBomby replied:
You're very right. I forgot the Ctrl+A loads every single selection blueprint. I'll find some way to get only the visible blueprints.

The performance impact is probably small because it only does a simple operation per each blueprint.

OliBomby committed:
only consider hit objects which are visible in timeline for contracted state

bdach committed:
Remove hide sample toggle

In line with feedback from
https://github.com/ppy/osu/pull/29896#issuecomment-2378676851.

bdach committed:
Merge branch 'master' into hs-toggle2

bdach commented:
I'm not sure why neither https://github.com/ppy/osu/pull/29896#issuecomment-2363054820 nor https://github.com/ppy/osu/pull/29896#issuecomment-2378676851 were ever responded to, but to move this forward I removed the toggle in https://github.com/ppy/osu/pull/29896/commits/8f48682d0a8a8385d2ade6db7ce0f03eac1668b4. We can revisit if the collapse change is deemed insufficient. (A pesky merge conflict gave it away.)

Otherwise seems fine, but I will wait for the @peppy UX sign-off.

OliBomby commented:
Sorry for not responding to the comments. I'm fine with going the current direction and revisiting the toggle if this is deemed insufficient.

peppy committed:
Merge branch 'master' into hs-toggle2
